src/Database/LSMTree/General.hs

```diff
@@ -0,0 +1,295 @@
+{-# LANGUAGE GADTs                    #-}
+{-# LANGUAGE RankNTypes               #-}
+{-# LANGUAGE StandaloneKindSignatures #-}
+
+-- TODO: remove once the API is implemented.
+{-# OPTIONS_GHC -Wno-redundant-constraints #-}
+{-# OPTIONS_GHC -Wno-unused-top-binds #-}
+
+module Database.LSMTree.General (
+    -- * Instantiations
+    NormalTableHandle
+  , MonoidalTableHandle
+  , NormalOp
+  , MonoidalOp
+    -- * Utility types
+  , IOLike
+    -- * Table sessions
+  , TableSession
+  , TableConfig
+  , newTableSession
+  , closeTableSession
+    -- * Tables
+  , TableHandle
+  , newTable
+  , openTable
+  , closeTable
+    -- * Table querying and updates
+    -- ** Queries
+  , Range
+  , lookups
+  , rangeLookup
+    -- ** Updates
+  , submitOps
+    -- ** Blobs
+  , BlobRef
+  , retrieveBlobs
+    -- * Snapshots
+  , VerificationFailure
+  , SnapshotName
+  , takeSnapshot
+  , loadSnapshot
+    -- * Multiple writable table handles
+  , duplicate
+    -- * Merging tables
+  , mergeTables
+  ) where
+
+import           Control.Concurrent.Class.MonadMVar (MonadMVar)
+import           Control.Monad.Class.MonadThrow (MonadCatch, MonadThrow)
+import           Data.Kind (Type)
+import           Data.Void
+import           System.FS.API (FsPath, HasFS, SomeHasFS)
+
+{-------------------------------------------------------------------------------
+  Instantiations
+-------------------------------------------------------------------------------}
+
+type NormalTableHandle   m k v blob = TableHandle m k v blob NormalOp
+type MonoidalTableHandle m k v      = TableHandle m k v Void MonoidalOp
+
+data NormalOp v blob =
+    NInsert !v !(Maybe blob)
+  | NDelete
+
+data MonoidalOp v blob =
```

Comment:
I'm not sure I understand how it can be so general. How will the `TableHandle` functions deal with any `op`?

Reply:
With appropriate constraints, I supposed it would be possible. With something like (crude example):

```haskell
class OpConstraints op v blob where
  -- | Merge/resolve operations. Should be associative
  combineOps :: op v blob -> op v blob -> op v blob
  -- | Is it safe to remove this operation from the last level of an LSM tree? 
  --
  -- Example: wrt inserts and deletes, deletes can be removed safely from
  -- the last level
  safeToRemove :: op v blob -> Bool
  serialiseOp :: op v blob -> Encoding
  deserialiseOp :: Decoder s (op v blob) 
```

Reply:
Not applicable anymore